New quickstart (#312)
zeek/zeek-docs: get-started/quickstart.rst

.. _ZeekControl documentation: https://github.com/zeek/zeekctl
.. _FAQ: https://zeek.org/faq/
.. _Get Zeek: https://zeek.org/get-zeek/
.. _Zeek source code: https://github.com/zeek/zeek
.. _gzip: https://www.gzip.org/

.. _quickstart:

=================
Quick Start Guide
=================

Zeek is a network traffic analyzer. Zeek works on most modern Unix-based
systems and does not require custom hardware. See :doc:`install` in order to
install from pre-built binary packages, or :doc:`building-from-source` in order
to build Zeek from source.

> Zeek requires some network traffic in order to run. For this guide, we’ll use

Author: Changed it a little bit from the suggestion but no longer spreading lies :)

created conn.log has two entries. Here are those entries (some fields are cut
for brevity with "..."):

   .. code-block:: console

     {
        "ts": 1747147647.668533,
        "uid": "CmgO6f3ddanrRjiCoc",
        ...
     },
     {
        "ts": 1747147654.27566,
        "uid": "CQuwSY2FD0NsLn9CJj",
        ...
     }

Now to find out more about both of these connections. Looking at the HTTP log,
it also has two entries. Match these entries with the connection log using the
timestamp and unique identifier (UID). The abbreviated logs for these two
requests are:

   .. code-block:: console

     {
         "ts": 1747147647.702181,
         "uid": "CmgO6f3ddanrRjiCoc",
         "method": "GET",
         "host": "zeek.org",
         "uri": "/",
         ...
     },
     {
         "ts": 1747147654.311012,
         "uid": "CQuwSY2FD0NsLn9CJj",
         "method": "WEIRD",
         "host": "zeek.org",
         "uri": "/",
         ...
     }

The first entry is a simple ``GET`` request to zeek.org. The second entry has
the same UID as the first entry in conn.log. The HTTP method was ``WEIRD``. For
this case, Zeek will also generate a separate log for some unexpected behavior

> (or "weirds"), like an unknown HTTP method. You can find this in weird.log:

Reviewer: > I just liked the word play.

Then you should keep it `WEIRD` :-)

   .. code-block:: console

     {
       "ts": 1747147654.311012,
       "uid": "CQuwSY2FD0NsLn9CJj",
       "name": "unknown_HTTP_method",
       "addl": "WEIRD",
       ...
     }

The UID for this entry is the same as the first entry in conn.log and the second
entry in http.log. Therefore, there were two HTTP requests, one with a ``GET``
request and one with a ``WEIRD`` request. The ``WEIRD`` request was rightfully
classified as a “weird” by Zeek.

More information on the various logs and what they report can be found in the
:doc:`../logs/index` section. More information on working with logs can be found in
the :doc:`../log-formats` section.

Live Traffic
============

Zeek is often run on live traffic on a network, not just captured traffic.
You can provide Zeek with a network interface to monitor traffic. Any traffic on
that interface will be analyzed in order to create logs. For example, you may
run Zeek on the ``en0`` network device (change ``en0`` to the device you want
to monitor traffic on):

.. code-block:: console

>      $ zeek -i en0

Author: I'll use this as an opportunity to add a `note` - I like how it breaks up the page and it adds a nice explanation, I think.

Then, in another terminal, create the same two HTTP requests we saw earlier via
``curl``:

.. code-block:: console

     $ curl -X GET http://zeek.org
     $ curl -X WEIRD http://zeek.org

Zeek will analyze traffic over that network device. Therefore, the logs may
have more than just the two entries found before. Entries should still appear in
:file:`conn.log`, :file:`http.log`, and :file:`weird.log` for these commands.

Scripting
=========

> Zeek can also use its own scripting language in order to configure behavior and

Reviewer: ```suggestion
You can also use Zeek's own scripting language in order to modify and extend its behavior:
```
If this is totally off from how you want this to sound, reject. Just seemed the user would be the one using the scripting language.

react to events:

.. code-block:: zeek

     event http_request(c: connection, method: string, original_URI: string,
         unescaped_URI: string, version: string)
         {
         print fmt("HTTP request: %s %s (%s->%s)", method, original_URI, c$id$orig_h,
             c$id$resp_h);
         }

This script defines an event that will run whenever Zeek sees an HTTP request.
You can run it through Zeek with the data from the pcap:

.. code-block:: console

     $ zeek weird-test.zeek -r quickstart.pcap
     HTTP request: GET / (192.168.1.8->192.0.78.212)
     HTTP request: WEIRD / (192.168.1.8->192.0.78.212)

Or on live traffic:

.. code-block:: console

>      $ zeek -i en0

Reviewer: ```suggestion
     $ zeek weird-test.zeek -i en0
```

In another terminal, run the two ``curl`` commands from before:

.. code-block:: console

     $ curl -X GET http://zeek.org
     $ curl -X WEIRD http://zeek.org

The terminal running Zeek will print each command as it gets processed.

More information on how to use Zeek’s scripting language can be found in the
:doc:`../scripting/index` section.

Managing Zeek
=============

> Zeek comes packaged with ZeekControl (``zeekctl``) to manage more complex use

Reviewer: ```suggestion
Zeek comes packaged with ZeekControl (``zeekctl``) to manage more complex deployments.
```

cases.

The same network device used in the Zeek command line can be used with
``zeekctl``. This will go in a configuration file. For the following example,
``$PREFIX`` will refer to the installation directory. This is likely
``/usr/local/zeek`` if built from source or ``/opt/zeek`` if installed from a
pre-built package.

First, update the configuration’s interface in ``$PREFIX/etc/node.cfg``. If
the device is ``en0``, that would look like:

.. code-block:: console

     [zeek]
     type=standalone
     host=localhost
     interface=en0

> You can also further configure the ``local.zeek`` script found in

Reviewer: ```suggestion
You can further configure the ``local.zeek`` script found in
```

> ``$PREFIX/share/zeek/site/local.zeek``. This script is loaded by ``zeekctl`` by

Reviewer: ```suggestion
``$PREFIX/share/zeek/site/local.zeek``. ``zeekctl`` loads this script by default.
```

Author: I seem to have a fear of starting sentences with lowercase command names :)

Since your suggestion makes the sentence in active voice, though, I think it's better.

default. It is not overwritten by Zeek upgrades. Zeek without ``zeekctl`` may

> specify ``local.zeek`` in the command line in order to load its contents.

Reviewer: While `local.zeek` not being added to a bare `zeek` invocation is definitely a known gotcha, unclear sneaking it in here is useful - maybe leave it out?

Author: you're right - I think just waiting until this point to mention it might be enough context without adding an extra explanation

> Run ``zeekctl`` in order to get into a new prompt:

Reviewer: ```suggestion
Run ``zeekctl`` in order to start an interactive prompt to manage your Zeek deployment:
```

`ZeekControl documentation`_. More information on setting up a cluster can be
found in the :doc:`cluster-setup` section.

Clusters
========

ZeekControl is also used to manage a cluster of Zeek processes. This cluster
contains many processes which analyze traffic together. Clusters are normally
split among many machines. For this example, all nodes will be local.

First, return to the ``$PREFIX/etc/node.cfg`` configuration file. It currently
contains one "standalone" node: 

.. code-block:: console

     [zeek]
     type=standalone
     host=localhost
     interface=en0

A standalone node is not in a cluster. Instead, this will change to multiple
nodes which work together. The following configuration is commented out in the
``node.cfg`` file by default. Remove the ``[zeek]`` node from above and paste
this into the file (or uncomment the code):

.. code-block:: console

     [logger]
     type=logger
     host=localhost

     [manager]
     type=manager
     host=localhost

     [proxy]
     type=proxy
     host=localhost

     [worker-1]
     type=worker
     host=localhost
     interface=en0

     [worker-2]
     type=worker
     host=localhost

>      interface=en0

Reviewer: Just using a single worker might be better in quickstart. There's too many caveats with multiple workers (proper, load-balancing, af_packet_fanout_id, ...). I'm even a bit worried that running through zeekctl is overwhelming, but there's also no real way around it for a real deployment.

Author: I agree, I only kept that one because it's commented out by default, so I don't need to tell people to only comment one out. That could cause frustration if they are only *kind of* following the guide.

Maybe I just remove mentions of uncommenting altogether - but then people might see them anyway. I'll grapple with this in my head for a bit to see if there's something cleaner, or maybe those wouldn't be issues?

.. _ZeekControl documentation: https://github.com/zeek/zeekctl
.. _FAQ: https://zeek.org/faq/
.. _Get Zeek: https://zeek.org/get-zeek/
.. _Zeek source code: https://github.com/zeek/zeek
.. _gzip: https://www.gzip.org/

.. _quickstart:

=================
Quick Start Guide
=================

Zeek is a network traffic analyzer. Zeek works on most modern Unix-based
systems and does not require custom hardware. See :doc:`install` in order to
install from pre-built binary packages, or :doc:`building-from-source` in order
to build Zeek from source.

We will first analyze previously captured network traffic from a ``pcap`` file -
:download:`quickstart.pcap`. Later, we will use Zeek to monitor live traffic.

Running Zeek
============

Open a terminal, navigate to a clean directory, and run zeek on the
:file:`quickstart.pcap` file:

   .. code-block:: console

>      zeek -r quickstart.pcap LogAscii::use_json=T

Author: I added a note at the end of the section, since other sections don't actually use that. Do you like that? I don't want to add it every time, but when showing logs I *really* wanted to use JSON.

Also I don't want to take it for granted that they'd know. I think it should be explained somewhere

>      zeek -r quickstart.pcap LogAscii::use_json=T

Reviewer 2: Yep, that looks good to me.

Scripting
=========

You can also use Zeek's own scripting language in order to modify and extend
its behavior:

.. code-block:: zeek

     # example.zeek
     event http_request(c: connection, method: string, original_URI: string,
         unescaped_URI: string, version: string)
         {
         print fmt("HTTP request: %s %s (%s->%s)", method, original_URI, c$id$orig_h,
             c$id$resp_h);
         }

This script defines an event handler that will run whenever Zeek sees an HTTP
request. You can run it through Zeek with the data from the pcap:

.. code-block:: console

     $ zeek example.zeek -r quickstart.pcap
     HTTP request: GET / (192.168.1.8->192.0.78.212)
     HTTP request: WEIRD / (192.168.1.8->192.0.78.212)

Or on live traffic:

.. code-block:: console

     $ zeek example.zeek -i en0 -C

In another terminal, run the two ``curl`` commands from before:

.. code-block:: console

     $ curl -X GET http://zeek.org
     $ curl -X WEIRD http://zeek.org

The terminal running Zeek will print each command as it gets processed.

More information on how to use Zeek’s scripting language can be found in the
:doc:`../scripting/index` section.

Managing Zeek
=============

> Zeek comes packaged with ZeekControl (``zeekctl``) to manage more complex

Author: I have a link at the end, the formula I was using is "show first, then link away at the end of the section" - which I prefer here so that there's an initial explanation. Not super hard to throw another link here, but I'd actually prefer not

     [ZeekControl] > stop
     stopping zeek ...
     [ZeekControl] > exit

And exit from ``zeekctl``:

.. code-block:: console

     [ZeekControl] > exit

The logs from ZeekControl will not appear in your current directory. Instead,
they will appear in ``$PREFIX/logs/current`` when running. Since the process was
stopped, they will appear in a directory with the current date within 
``$PREFIX/logs/`` - such as ``$PREFIX/logs/2025-01-01/``.

These logs are compressed as ``.log.gz`` files from gzip_. You may decompress
these via ``gunzip`` then read them, or use gzip’s packaged ``zcat`` command.
On Mac, this looks like:

.. code-block:: console

     $ zcat < $PREFIX/logs/2025-01-08/weird.11:03:38-11:03:43.log.gz
     <...>
     1736352218.157077       CFvENWVlkwVHhLL35       2603:6081:18f0:99e0:7da2:6b81:9a83:cb4e 57823   2606:2800:21f:cb07:6820:80da:af6b:8b2c   80      unknown_HTTP_method     WEIRD   F       zeek    -

The logs contain the ``WEIRD`` HTTP request.

More information on using ZeekControl can be found in the
`ZeekControl documentation`_. More information on setting up a cluster can be
found in the :doc:`cluster-setup` section.

Clusters
========

> ZeekControl is also used to manage a cluster of Zeek processes. This cluster

Reviewer 2: ```suggestion
ZeekControl is also used to manage a cluster of Zeek processes. A cluster
```

contains many processes which analyze traffic together. Clusters are normally

> split among many machines. For this example, all nodes will be local.

Reviewer 2: > normally split among many machines

I'm not sure this is actually true. I think more people run a single host with multiple processes than multi-host clusters.

Author: I reworded it to not spread falsities :)
